docs: add Oso blog post (#14905)
pulumi/docs: content/blog/how-we-used-pulumi-to-safely-migrate-osos-global-infrastructure/index.md

---
title: "How We Used Pulumi to Safely Migrate Oso's Global Infrastructure"
allow_long_title: true
date: 2025-05-13
draft: false

meta_desc: "Learn how Oso used Pulumi to safely migrate its global infrastructure, ensuring zero downtime and maximum leverage."
meta_image: meta.png

authors:
    - shaun-verch

tags:

>     - change-me

Reviewer: ```suggestion
    - aws
    - refactoring
    - infrastructure-as-code
```

social:
    twitter:
    linkedin:

---

Infrastructure as Code (IaC) tools such as Pulumi can provide enormous amounts of leverage, but they must be used correctly to also provide safety. One of our main jobs as infrastructure engineers is to not break things, so leverage without safety is useless. If something is safe, we can change things easily without even thinking about it. If it isn't, we'll be up until 2 a.m. fixing what we broke.

At Oso, we recently had to do a large infrastructure migration and learned three key principles that make a huge difference in using Pulumi for safe migrations. First, understand how it's even possible to safely change your infrastructure independently of Pulumi. Second, write reusable and predictable Pulumi code. Finally, refactor your Pulumi code to your new version using zero-diff refactors.

## Osoʼs Migration

![img.png](img.png)

> Oso provides Authorization as a Service, so every application decision must first query Oso. This means that if Oso goes down, our customers will also go down. We can't afford a moment of downtime.

Reviewer: ```suggestion
[Oso](https://www.osohq.com/) provides Authorization as a Service, so every application decision must first query Oso. This means that if Oso goes down, our customers will also go down. We can't afford a moment of downtime.
```

The holy grail of infrastructure development is to achieve what we already have in application development: code reuse. Besides the benefit of reducing the burden of maintaining more code, it also reduces risk because we can concentrate our testing in one place and benefit from using code that has already been proven.

In Osoʼs case, because our migration involved creating a copy of our entire infrastructure with the new changes, code reuse was vital. Even the smallest reusable component meant fewer new bugs and less time fixing bugs we had already fixed in the old infrastructure.

In Pulumi, you can use the import and function system of the underlying programming language, so that part of managing libraries is not an issue. The real challenge lies in how to actually write the business logic of libraries in Pulumi because writing Pulumi code is not like writing normal code.

Take a look at this example of creating an ECS service:

```typescript

let service = new aws.ecs.Serice("api",{ });

```

This line of code might seem trivial, but a lot is happening here, and it's important to understand Pulumi's programming model on a deeper level to know how to work with it properly.

> The first thing to note is the first argument: `api`. This is the logical resource.
> 
> name in Pulumi. The subsequent arguments are more about how to actually create the resource, and we won't talk about that here. This logical resource is how Pulumi identifies this resource in the internal state.

Reviewer 2: ```suggestion
The first thing to note is the first argument: ”api”. This is the [logical resource name](https://www.pulumi.com/docs/iac/concepts/resources/names/) in Pulumi. The subsequent arguments are more about how to actually create the resource, and we wonʼt talk about that here. This logical resource is how Pulumi identifies this resource in the internal state.
```

The internal state is the core of the Pulumi engine. It maps the “logical resource name" we provide to the actual resource that Pulumi either created or plans to create. The code example above, at some point deep inside the Pulumi engine, turns into something that roughly looks like this:

> ```json
> 
> {
> 
>   "aws:ecs:Service": {
> 
>     "api": {...},
> 
>     ...
> 
>   },
> 
>   ...
> 
> }
> 
> ```

Reviewer 2: ```suggestion
```json
{
  "aws:ecs:Service": {
    "api": {...},
    ...
  },
  ...
}
```
```

> This is the key to understanding why writing Pulumi code is not like writing normal code. There is one "state namespace" per Pulumi execution, which effectively means that there is one global object that we are adding resources to. So the `new aws.ecs.Service` function call wasn't just creating a service object that we can pass around; it was registering the to deploy. service with the global list of resources we plan

Reviewer 2: ```suggestion
This is the key to understanding why writing Pulumi code is not like writing normal code. There is one "state namespace" per Pulumi execution, which effectively means that there is one global object that we are adding resources to. So the `new aws.ecs.Service` function call wasn't just creating a service object that we can pass around; it was registering the service with the global list of resources we plan to deploy.
```
